runtime/src/chpl-privatization.c

```diff
@@ -19,16 +19,89 @@
 
 #include "chplrt.h"
 #include "chpl-privatization.h"
+#include "chpl-atomics.h"
 #include "chpl-mem.h"
 #include "chpl-tasks.h"
 
-static int64_t chpl_capPrivateObjects = 0;
-static chpl_sync_aux_t privatizationSync;
+#define CHPL_PRIVATIZATION_BLOCK_SIZE 1024
 
-void** chpl_privateObjects = NULL;
+static chpl_sync_aux_t writeLock;
+
+
+// Each block is an array of privatized objects
+typedef void ** chpl_priv_block_t;
+
+// Each instance is a variable array of blocks
+typedef struct {
+  chpl_priv_block_t * blocks;
+  size_t len;
+} chpl_priv_instance_t;
+
+// Manage 2 instances to switch between
+chpl_priv_instance_t chpl_priv_instances[2];
+
+// Reader Count for both instances. (MUST BE ATOMIC)
+atomic_uint_least32_t reader_count[2];
+
+// Determines current instance. (MUST BE ATOMIC)
+atomic_int_least8_t currentInstanceIdx;
+
+chpl_priv_block_t chpl_priv_block_create() {
```

Comment:
I never remember this, but in C 0 argument functions should take "void".  In fact, I just messed this up yesterday: https://github.com/chapel-lang/chapel/pull/8168

Reply:
Fixed, although I'm not 100% certain I understand the actual need (something about it being possible for someone to pass arguments to a no-argument function and mess with the stack contents... kinda interested in what happens in that case, actually...)

Reply:
For compatibility with old C, a no argument prototype actually means "Don't assume or check anything about the args", where's a void signature means "this is a 0 argument function" -- https://stackoverflow.com/questions/693788/is-it-better-to-use-c-void-arguments-void-foovoid-or-not-void-foo